main/src/library/ChannelImpl.flix

```diff
@@ -0,0 +1,379 @@
+        else Ok(unsafeNew(bufferSize))
+
+    ///
+    /// Creates a new channel. A runtime error occurs if `bufferSize` is negative.
+    ///
+    pub def unsafeNew(bufferSize: Int): ChannelImpl & Impure =
+        import dev.flix.runtime.GlobalCounter:newId();
+        let _bufferCheck = if (bufferSize < 0) inputBug("newChannelWithQueue", "bufferSize < 0") else ();
+        let buffered = bufferSize == 0;
+        let reentrantLock = ReentrantLock.new(false);
+        Mpmc(
+            newId(),
+            reentrantLock,
+            buffered,
+            if (buffered) 1 else bufferSize,
+            MutDeque.new(),
+            MutList.new(),
+            ReentrantLock.newCondition(reentrantLock)
+        )
+
+    ///
+    /// Sends the element `boxedElement` on the channel `channel`.
+    ///
+    /// Implements the expression `channel <- x`.
+    ///
+    pub def put(channel: ChannelImpl, boxedElement: Boxed): ChannelImpl & Impure =
+        let Mpmc(_, channelLock, buffered, _, deque, waitingGetters, waitingSetters) = channel;
+        ReentrantLock.lock(channelLock);
+
+        // Block until the channel is not full
+        awaitAvailableSpace(channel);
+
+        // Insert the new element
+        MutDeque.pushBack(boxedElement, deque);
+
+        // Signal waitingGetters that there is an element available
+        let signalLockConditionPair = lockConditionPair -> {
+                let (conditionLock, condition) = lockConditionPair;
+                ReentrantLock.lock(conditionLock);
+                signalCondition(condition);
+                unlockLock(conditionLock)
+            };
+        MutList.foreach(signalLockConditionPair, waitingGetters);
+
+        // Clear waitingGetters.
+        // If a waitingGetter does not receive an element, it will add itself again
+        MutList.clear!(waitingGetters);
+
+        // If the channel is unbuffered, wait for the element to be handed off before continuing
+        // TODO: Optimization possibility. This await be faster with a separate condition such that the hand-off signal does not
+        //     have to fight for the channel lock with the threads that actually want to insert an element.
+        if (buffered) awaitCondition(waitingSetters)
+        else ();
+
+        unlockLock(channelLock);
+        channel
+
+    ///
+    /// Wait for the channel to have available space.
+    /// The channel lock is expected to be held.
+    ///
+    def awaitAvailableSpace(channel: ChannelImpl): Unit & Impure =
+        let Mpmc(_, _, _, maxQueueSize, deque, _, waitingSetters) = channel;
+        if (MutDeque.size(deque) == maxQueueSize) {
+            awaitCondition(waitingSetters);
+            awaitAvailableSpace(channel)
+        } else {
+            ()
+        }
+
+    ///
+    /// Receives an element from the channel `channel`.
+    ///
+    /// Implements to the expression `<- channel`.
+    ///
+    pub def get(channel: ChannelImpl): Boxed & Impure =
```

Comment:
I would put the public methods nearer the top and order them: new, get, put, select.

Reply:
channel -> c